File: ch05/PriorityQueue.java
package ch05;

@SuppressWarnings("unchecked")
public class PriorityQueue<T extends Comparable<T>> {
    /**
     * 当前存的节点容量大小
     */
    private int size;

    /**
     * 最大的容量大小
     */
    private int maxSize;

    /**
     * 数据存储地方
     */
    private Object[] values;

    public PriorityQueue(int maxSize) {
        this.maxSize = maxSize;
        this.size = 0;
        this.values = new Object[maxSize];
    }

    private T get(int i) {
        return (T) values[i];
    }

    /**
     * 获取最大值
     *
     * @return
     */
    public T peek() {
        return (T) values[1];
    }

    /**
     * 插入新结点
     */
    public void push(T t) {
        if (size == maxSize) {
            throw new UnsupportedOperationException("堆已满");
        }

        // 将新结点插到树最后的位置
        values[0] = values[++size] = t;
        // 上滤调整
        percolateUp();
    }

    /**
     * 上滤
     */
    protected void percolateUp() {
        for (int i = size; get(i / 2).compareTo(get(0)) < 0; i = i / 2) {
            Object node = values[i / 2];
            values[i / 2] = values[i];
            values[i] = node;
        }
    }

    /**
     * 删除最大结点
     */
    public T pop() {
        if (size == 0) {
            throw new UnsupportedOperationException("无法删除空树");
        }

        Object node = values[1];
        values[0] = values[1] = values[size--];
        percolateDown(1);

        return (T) node;
    }

    /**
     * 下滤
     */
    protected void percolateDown(int parent) {
        // 如果左孩子都不存在的话, 那就是叶子结点了, 因为这是一棵完全二叉树
        while (parent * 2 <= size) {
            int pos = parent * 2;
            int rightPos = parent * 2 + 1;

            // 如果右子树存在并且右 > 左, 那 pos = 右, 否则 pos = 左
            if (rightPos <= size && get(rightPos).compareTo(get(pos)) > 0) {
                pos = rightPos;
            }

            // 比较孩子结点和父结点的大小, pos 可能是左也可能是右
            // 不大于则说明最大结点就是父结点
            if (get(parent).compareTo(get(pos)) < 0) {
                Object tmp = values[parent];
                values[parent] = values[pos];
                values[pos] = tmp;
                // 转到下一层
                parent = pos;
            } else {
                // 父结点是最大的就退出比较, 否则交换位置
                break;
            }
        }
    }

    public static <T extends Comparable<T>> PriorityQueue<T> createQueue(T... arr) {
        PriorityQueue<T> queue = new PriorityQueue<>(arr.length * 2);
        queue.size = arr.length;
        System.arraycopy(arr, 0, queue.values, 1, arr.length);

        for (int i = queue.size / 2; i > 0; i--) {
            queue.percolateDown(i);
        }

        return queue;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder(size);
        builder.append("[");
        boolean flag = false;
        for (int i = 1; i <= size; i++) {
            if (flag) {
                builder.append(", ");
            } else {
                flag = true;
            }
            builder.append(values[i]);
        }

        return builder.append("]").toString();
    }

    public static void main(String[] args) {
        Integer[] arr = { 79, 66, 16, 83, 30, 19, 68, 55, 91, 72, 49, 9 };
        PriorityQueue<Integer> queue = new PriorityQueue<>(20);

        System.out.println("插入数据演示:");
        for (int value : arr) {
            System.out.println("插入新数据: " + value);
            queue.push(value);
        }

        System.out.println("生成的堆: " + queue);
        System.out.println("---------------------------------------\n");

        System.out.println("删除数据演示:");
        while (queue.size > 0) {
            Integer value = queue.pop();
            System.out.println("删除数据: " + value);
        }

        System.out.println("---------------------------------------\n");

        queue = PriorityQueue.createQueue(arr);
        System.out.println("O(N) 时间复杂度生成一个新堆: " + queue);
    }
}

File: ch05/priority_queue.go
package ch05

import (
	"errors"
	"fmt"

	"github.com/ivicel/zju_data_structure/ch03"
)

type EqualFunc func(*ch03.TreeNode, *ch03.TreeNode) int

// 最大堆
type PriorityQueue struct {
	Size    int             // 当前队列大小
	MaxSize int             // 队列的最大容量
	Data    []ch03.TreeNode // 存储数据结点
	Equal   EqualFunc
}

// 插入新的结点
func (q *PriorityQueue) Push(val interface{}) error {
	if q.Size == q.MaxSize {
		return errors.New("队列已满, 插入失败")
	}

	// 将结点放到最后位置
	newNode := ch03.TreeNode{Data: val}
	q.Size++
	q.Data[q.Size] = newNode
	// 上滤
	pos := q.Size
	for q.Data[0] = newNode; q.Equal(&newNode, &q.Data[pos/2]) > 0; pos = pos / 2 {
		// 大于时交换值
		q.Data[pos], q.Data[pos/2] = q.Data[pos/2], q.Data[pos]
	}

	return nil
}

// 删除最大值
func (q *PriorityQueue) Pop() (ch03.TreeNode, error) {
	if q.Size == 0 {
		return ch03.TreeNode{}, errors.New("不能删除空树")
	}

	// 保存根结点
	node := q.Data[1]
	q.Data[1] = q.Data[q.Size]
	q.Size--

	q.percolateDown(1)

	return node, nil
}

// 下滤
func (q *PriorityQueue) percolateDown(root int) {
	q.Data[0] = q.Data[root]
	// 如果左孩子都不存在的话, 那就是叶子结点了, 因为这是一棵完全二叉树
	for root*2 <= q.Size {
		pos, rightPos := root*2, root*2+1

		// 如果右子树存在并且右 > 左, 那 pos = 右, 否则 pos = 左
		if rightPos <= q.Size && q.Equal(&q.Data[pos], &q.Data[rightPos]) < 0 {
			pos = rightPos
		}

		// 比较孩子结点和父结点的大小, pos 可能是左也可能是右
		// 不大于则说明最大结点就是父结点
		if q.Equal(&q.Data[pos], &q.Data[0]) > 0 {
			q.Data[root] = q.Data[pos]
			root = pos
		} else {
			// 最大值就是父结点时, 不用交换
			break
		}
	}
	q.Data[root] = q.Data[0]
}

// 打印树内容
func (q *PriorityQueue) String() string {
	return fmt.Sprintf("%v", q.Data[1:q.Size+1])
}

// 初始化一个空堆
func InitQueue(maxSize int, cmp EqualFunc) *PriorityQueue {
	queue := &PriorityQueue{Size: 0, MaxSize: maxSize, Equal: cmp}
	queue.Data = make([]ch03.TreeNode, maxSize)
	return queue
}

// 根据传入的切片数据生成一个新堆
func CreateQueue(arr []interface{}, cmp EqualFunc) *PriorityQueue {
	// 按默认顺序创建一个新堆
	size := len(arr)
	maxSize := size * 2
	queue := InitQueue(maxSize, cmp)

	nodes := make([]ch03.TreeNode, size, maxSize)
	for i, val := range arr {
		nodes[i] = ch03.TreeNode{Data: val}
	}
	queue.Data = nodes

	// 调整排序
	for i := size / 2; i > 0; i-- {
		queue.percolateDown(i)
	}

	return queue
}
